applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/differentiation_utility.cpp

```diff
@@ -0,0 +1,258 @@
+            if ( (rOutput.size1() != 0) || (rOutput.size2() != 0) )
+                rOutput.resize(0,0,false);
+        }
+
+        KRATOS_CATCH("");
+    }
+
+    void DifferentiationUtility::CalculateMassMatrixDerivative(Element& rElement,
+                                                const array_1d_component_type& rDesignVariable,
+                                                Node<3>& rNode,
+                                                const double& rPertubationSize,
+                                                Matrix& rOutput,
+                                                ProcessInfo& rCurrentProcessInfo)
+    {
+        KRATOS_TRY;
+
+        #ifdef _OPENMP
+            KRATOS_ERROR_IF(omp_get_thread_num() > 0) <<
+                "DifferentiationUtility::CalculateMassMatrixDerivative " <<
+                "is not thread safe for shape derivatives!" << omp_get_thread_num();
+        #endif
+
+        if( rDesignVariable == SHAPE_X || rDesignVariable == SHAPE_Y || rDesignVariable == SHAPE_Z )
+        {
+            const IndexType coord_dir = DifferentiationUtility::GetCoordinateDirection(rDesignVariable);
+
+            // define working variables
+            Matrix unperturbed_mass_matrix;
+            Matrix perturbed_mass_matrix;
+
+            // compute mass matrix before perturbion
+            rElement.CalculateMassMatrix(unperturbed_mass_matrix, rCurrentProcessInfo);
+
+            if ( (rOutput.size1() != unperturbed_mass_matrix.size1()) || (rOutput.size2() != unperturbed_mass_matrix.size2() ) )
+                rOutput.resize(unperturbed_mass_matrix.size1(), unperturbed_mass_matrix.size2());
+
+            // perturb the design variable
+            rNode.GetInitialPosition()[coord_dir] += rPertubationSize;
+            rNode.Coordinates()[coord_dir] += rPertubationSize;
+
+            // compute LHS after perturbation
+            rElement.CalculateMassMatrix(perturbed_mass_matrix, rCurrentProcessInfo);
+
+            //compute derivative of RHS w.r.t. design variable with finite differences
+            noalias(rOutput) = (perturbed_mass_matrix - unperturbed_mass_matrix) / rPertubationSize;
+
+             // unperturb the design variable
+            rNode.GetInitialPosition()[coord_dir] -= rPertubationSize;
+            rNode.Coordinates()[coord_dir] -= rPertubationSize;
+
+            //call one last time to make sure everything is as it was before TODO improve this..
+            rElement.CalculateMassMatrix(perturbed_mass_matrix, rCurrentProcessInfo);
+        }
+        else
+        {
+            KRATOS_WARNING("DifferentiationUtility") << "Unsupported nodal design variable: " << rDesignVariable << std::endl;
+            if ( (rOutput.size1() != 0) || (rOutput.size2() != 0) )
+                rOutput.resize(0,0,false);
+        }
+
+        KRATOS_CATCH("");
+    }
+
+    std::size_t DifferentiationUtility::GetCoordinateDirection(const array_1d_component_type& rDesignVariable)
+    {
```

Comment:
this function could be simplified to the following (just a suggestion):
~~~
        if( rDesignVariable == SHAPE_X )
            return 0;
        else if( rDesignVariable == SHAPE_Y )
            return 1;
        else if( rDesignVariable == SHAPE_Z )
            return 2;
        else
             Error
~~~